### PR TITLE
[MLIR][TORCH] Add E2E support for aten.max_pool2d_with_indices_backward op

### PR DESCRIPTION
**Pull request by erman-gurses**

### CODE REVIEW

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/upstream_shape_helpers.py [hunk 1]
```diff
@@ -216,6 +216,10 @@ def max_pool2d_with_indices(input: List[int], kernel_size: List[int], stride: Li
   out = max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
   return (out, out)
 
+def max_pool2d_with_indices_backward(input: List[int], kernel_size: List[int], stride: List[int], padding: List[int], dilation: List[int], ceil_mode: bool):
```
Comment (silvasean) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/upstream_shape_helpers.py:219]: This file should not be edited, except by copying upstream code into it. I just added a note into top fo the file: https://github.com/llvm/torch-mlir/commit/94df096c11011667ddd0da1a5058f7918732b5dc

**Review by silvasean — CHANGES_REQUESTED** [review]

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/upstream_shape_helpers.py (hunk 1 above)
Comment (silvasean) [reply python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/upstream_shape_helpers.py:219]: You can just put this code into the main shape_lib_gen.py

Comment (erman-gurses) [reply python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/upstream_shape_helpers.py:219]: Sure, I will follow that.

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 2]
```diff
@@ -302,6 +302,108 @@ class ConvertValsemVariantAtenIndexPutImplOp
...
+
+    //Value indicesList;
+    //if (!getListConstructElements(adaptor.indices(), indicesList))
+     // return rewriter.notifyMatchFailure(op, "indices not constructed from ListConstruct");
+    
+
+    ValueTensorType indexType =
+        adaptor.indices().getType().cast<ValueTensorType>();
+
+    Value expandedIndexSizes{indexType.getSizes(), 1};
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:377]: We initialize the vector this way. Change the `Value expanded...` -> `SmallVector<Value> expanded...`

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 3]
```diff
@@ -302,6 +302,108 @@ class ConvertValsemVariantAtenIndexPutImplOp
...
+      return rewriter.notifyMatchFailure(op, "unsupported dtype");
+    }
+
+    Value output = createZeroInitTensor(rewriter, loc, inputShape, inputEType);
+    //Value numTensorElements = rewriter.create<AtenNumelOp>(loc, input);
+    //rewriter.create<AtenZerosOp>(loc, inputType, numTensorElements, inputEType);
+
+    //Value indicesList;
+    //if (!getListConstructElements(adaptor.indices(), indicesList))
+     // return rewriter.notifyMatchFailure(op, "indices not constructed from ListConstruct");
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:371]: Could you please tell me the significance of this code? Whether you want this code or not?

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp (hunk 2 above)
Comment (erman-gurses) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp]: I updated those as we discussed.

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp (hunk 3 above)
Comment (erman-gurses) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp]: I was just doing trials. Not important ones now. Removed them.

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 4]
```diff
@@ -302,6 +302,91 @@ class ConvertValsemVariantAtenIndexPutImplOp
...
+    
+    //Expand indices
+    ValueTensorType indexType = op.indices().getType().cast<ValueTensorType>();
+    SmallVector<int64_t> expandedIndexSizes{indexType.getSizes()[0], 1};
+    ValueTensorType expandedIndexType = ValueTensorType::get(
+      context, llvm::makeArrayRef(expandedIndexSizes), indexType.getDtype());
+    Value torchCstOne = rewriter.create<Torch::ConstantIntOp>(
+      loc, rewriter.getI64IntegerAttr(1));
+    Value expandedIndexTensor = rewriter.create<AtenUnsqueezeOp>(
+      loc, expandedIndexType, adaptor.indices(), torchCstOne);
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:368]: adaptor.indices() -> op.indices()
The `AtenUnsqueezeOp` is a torch op and expects operands to be of the torch type.

**Pull request closed without merging**